File: engine/src/Player.ts
// 玩家：拥有金币 / 收入 / 已建建筑 / 策略回调
// 一个 BattleSim 持有两个 Player (Left/Right)

import type { Side } from "./Enums.ts";
import type { Building } from "./Building.ts";
import type { PlayerStrategy } from "./Strategy.ts";

export class Player {
  side: Side;
  race: string;
  name: string;

  gold: number;
  /** 下一次结算收入的绝对时间 (秒) */
  nextIncomeAt: number;
  ownedBuildings: Building[] = [];
  strategy: PlayerStrategy | null = null;

  /** 统计：总收入 / 总花费 / 造兵数 */
  stats = { totalIncome: 0, totalSpent: 0, unitsSpawned: 0 };

  constructor(side: Side, race: string, name: string, startGold: number, firstIncomeAt: number) {
    this.side = side;
    this.race = race;
    this.name = name;
    this.gold = startGold;
    this.nextIncomeAt = firstIncomeAt;
  }

  /** 当前每个收入周期的总收入 */
  incomePerInterval(baseIncome: number): number {
    let sum = baseIncome;
    for (const b of this.ownedBuildings) {
      if (!b.alive) continue;
      sum += b.tpl.incomeBonusPer10s ?? 0;
    }
    return sum;
  }

  /** 消费金币；不足返回 false */
  spend(cost: number): boolean {
    if (this.gold < cost) return false;
    this.gold -= cost;
    this.stats.totalSpent += cost;
    return true;
  }

  earn(n: number) {
    this.gold += n;
    this.stats.totalIncome += n;
  }
}

File: unity/Assets/Mg01/Sim/Player.cs
// 与 engine/src/Player.ts 一一对应

using System.Collections.Generic;
using Mg01.Core;
using Mg01.Templates;

namespace Mg01.Sim
{
    public class Player
    {
        public Side side;
        public string race;
        public string name;

        public int gold;
        public float nextIncomeAt;
        public List<Building> ownedBuildings = new();
        public PlayerStrategy strategy;

        public PlayerStats stats = new();

        public Player(Side s, string race, string name, int startGold, float firstIncomeAt)
        {
            this.side = s;
            this.race = race;
            this.name = name;
            this.gold = startGold;
            this.nextIncomeAt = firstIncomeAt;
        }

        public int IncomePerInterval(int baseIncome)
        {
            int sum = baseIncome;
            foreach (var b in ownedBuildings)
                if (b.alive) sum += b.tpl.incomeBonusPer10s;
            return sum;
        }

        public bool Spend(int cost)
        {
            if (gold < cost) return false;
            gold -= cost;
            stats.totalSpent += cost;
            return true;
        }

        public void Earn(int n)
        {
            gold += n;
            stats.totalIncome += n;
        }
    }

    public class PlayerStats { public int totalIncome, totalSpent, unitsSpawned; }

    /// <summary>策略接口；具体实现 (buildOrder / mixedComposition) 待移植自 engine/src/Strategy.ts</summary>
    public interface PlayerStrategy
    {
        string Decide(StrategyContext ctx);
    }

    public struct StrategyContext
    {
        public float now;
        public int gold, income;
        public int availableSlots;
        public IReadOnlyList<Building> ownedBuildings;
        public System.Func<string, bool> canAfford;
        public System.Func<string, int> ownedCountOf;
    }
}
